router/internal/httpclient/types.go

```diff
@@ -0,0 +1,144 @@
+package httpclient
+
+import (
+	"sort"
+	"time"
+)
+
+type DNSStart struct {
+	Time time.Time
+	Host string
+}
+
+type DNSDone struct {
+	Time      time.Time
+	Coalesced bool
+	Error     error
+}
+
+type TLSStart struct {
+	Time time.Time
+}
+
+type TLSDone struct {
+	Time      time.Time
+	Complete  bool
+	DidResume bool
+	Error     error
+}
+
+type DialStart struct {
+	Time    time.Time
+	Network string
+	Address string
+}
+
+type DialDone struct {
+	Time    time.Time
+	Network string
+	Address string
+	Error   error
+}
+
+type WroteHeaders struct {
+	Time time.Time
+}
+
+type WroteRequest struct {
+	Time  time.Time
+	Error error
+}
+
+type FirstByte struct {
+	Time time.Time
+}
+
+type AcquiredConnection struct {
+	Time     time.Time
+	Reused   bool
+	WasIdle  bool
+	IdleTime time.Duration
+}
+
+type GetConnection struct {
+	Time     time.Time
+	HostPort string
+}
+
+type ClientTrace struct {
+	ConnectionGet      *GetConnection
+	ConnectionAcquired *AcquiredConnection
+	DNSStart           *DNSStart
+	DNSDone            *DNSDone
+	TLSStart           *TLSStart
+	TLSDone            *TLSDone
+	DialStart          []DialStart
+	DialDone           []DialDone
+	WroteHeaders       *WroteHeaders
+	WroteRequest       *WroteRequest
+	FirstByte          *FirstByte
+}
+
+type ClientTraceInfo struct {
+	ClientTraces          []*ClientTrace
+	RetryCountLatestIndex int
+}
+
+type DialCombined struct {
+	DialStartTime time.Time
+	DialDoneTime  *time.Time
+	Error         error
+	Network       string
+	Address       string
+}
+
+// GetGroupedDials returns a slice of DialCombined structs
+// Which are combined based on the network (tcp/udp/...) and address
+// We sort by the first dial that was successful, so that the most likely
+// used dial is first in the list
+func (r ClientTrace) GetGroupedDials() []DialCombined {
```

Comment:
This can be simplified. We can take the first DialStart event and ConnectStart to calculate the duration.

```go
var (
        mu            sync.Mutex
        startAll      time.Time
    )

    trace := &httptrace.ClientTrace{
        ConnectStart: func(_, _ string) {
            mu.Lock()
            if startAll.IsZero() {
                startAll = time.Now()
            }
            mu.Unlock()
        },
        GotConn: func(info httptrace.GotConnInfo) {
              mu.Lock()
              total := time.Since(startAll)
              mu.Unlock()
              fmt.Printf("Full connect (TCP+TLS) took %v\n", total)
        },
    }
```

Reply:
While we cannot use this, because we need to get the first **successful** "ConnectDone" that finished first. We could simplify it a bit by only keeping a map for the ConnectStart, and within this function itself. Which I will do.

Reply:
Also when `GotConn` is called after we successfully connected?